source/base/tensor_product_polynomials_bubbles.cc

```diff
@@ -0,0 +1,251 @@
+// ---------------------------------------------------------------------
+// $Id$
+//
+// Copyright (C) 2012 - 2013 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+
+#include <deal.II/base/tensor_product_polynomials_bubbles.h>
+#include <deal.II/base/exceptions.h>
+#include <deal.II/base/table.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+
+
+/* ------------------- TensorProductPolynomialsBubbles -------------- */
+
+
+template <int dim>
+double
+TensorProductPolynomialsBubbles<dim>::compute_value (const unsigned int i,
+                                                     const Point<dim> &p) const
+{
+  const unsigned int q_degree = this->polynomials.size()-1;
+  const unsigned int max_indices = this->n_tensor_pols;
+  const unsigned int n_bubbles = ((q_degree<=1)?1:dim);
+  Assert (i<max_indices+n_bubbles, ExcInternalError());
+
+  // treat the regular basis functions
+  if (i<max_indices)
+    return this->TensorProductPolynomials<dim>::compute_value(i,p);
+
+  const unsigned int comp = i - this->n_tensor_pols;
+
+  //compute \prod_{i=1}^d (1-x_i^2)(p)
+  double value=1.;
+  for (unsigned int j=0; j < dim; ++j)
+    value*=4*p(j)*(1-p(j));
+  // and multiply with x_i^{r-1}
+  value*=std::pow(2*p(comp)-1,q_degree-1);
```

Comment:
Since polynomials might get evaluated relatively often (e.g. when generating restriction matrices, when getting interface constraints, etc.), I'd prefer not to use std::pow() but rather write out a loop. Since we expect q_degree to be modest (say at most 5 or 6), a plain loop should be the fastest option rather than the optimal `log(q_degree)` implementation. (Or does std::pow actually use integer optimizations, then ignore this comment?)